toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1355,23 +1449,33 @@ func (g *PkgGraph) fixPrebuiltSRPMsCycle(trimmedCycle []*PkgNode) (err error) {
 
 			logger.Log.Debugf("Adding a 'PreBuilt' node '%s' with id %d.", preBuiltNode.FriendlyName(), preBuiltNode.ID())
 
-			parentNodes := g.To(currentNode.ID())
-			for parentNodes.Next() {
-				parentNode := parentNodes.Node().(*PkgNode)
-				if parentNode.Type == TypeLocalBuild && parentNode.SrpmPath == previousNode.SrpmPath {
-					g.RemoveEdge(parentNode.ID(), currentNode.ID())
-
-					err = g.AddEdge(parentNode, preBuiltNode)
-					if err != nil {
-						logger.Log.Errorf("Adding edge failed for %v -> %v", parentNode, preBuiltNode)
-						return
-					}
-				}
-			}
+			g.replaceCurrentNodeWithPrebuiltNode(currentNode, preBuiltNode, previousNode)
+			logger.Log.Infof("Cycle fixed using prebuilt node: %s with id %d", preBuiltNode.FriendlyName(), preBuiltNode.ID())
 
 			return
 		}
-
+		if resolveCyclesFromUpstream {
+			isAvailableUpstream := false
+			var clonerErr error
+			isAvailableUpstream, clonerErr = isSRPMAvailableUpstream(g, nil, currentNode.SrpmPath, ignoreVersionToResolveSelfDep, cloner)
```

Comment:
Would it be better to try to resolve all of the edges first with `ignoreVersionToResolveSelfDep` forced to false first, and only if we can't validate ANY of the edges optionally turn it on? That way we will check all the edges hoping to find an exact version match to one of the packages.

Reply:
This is already implemented inside ### findAllRPMSUpstream. It tries to find exact version match first. If not found and ### ignoreVersionToResolveSelfDep is true, then only we clear the version and search.

Reply:
@dmcilvaney Yes. It's a better way. I've updated the logic.